### DISCUSSION TITLE
$(tty) returning wrong TTY

### DISCUSSION THREAD

**Discussion by TobiasDeBruijn** [microsoft/WSL#6195] [General]
# Environment
```none
Windows build number: Microsoft Windows [Version 10.0.18363.1082]
Your Distribution version: 18.04
Whether the issue is on WSL 2 and/or WSL 1: Linux version 4.4.0-18362-Microsoft (Microsoft@Microsoft.com) (gcc version 5.4.0 (GCC) ) #1049-Microsoft Thu Aug 14 12:01:00 PST 2020
```

# Steps to reproduce

1. Create a ~/.bashrc file, containing ``export GPG_TTY = "$(tty)"``
2. Close terminal
3. Open a new terminal, do nothing with it.
4. Open another new terminal, run e.g ``git commit -S`` to prompt te GPG password prompt

#  Expected behavior
GPG Password prompt opens on the terminal where ``git commit -S`` was run from

# Actual behavior
GPG Password prompt opens on the terminal which was opened first

![image](https://user-images.githubusercontent.com/28569170/98110330-369fae80-1e9f-11eb-867a-93f6d9ef94ad.png)

# Possible explanation
I think the $(tty) in ~/.bashrc does not return the TTY of the terminal that was opened, but rather the terminal which was opened first, which would explain why the GPG password prompt opens on the wrong terminal.

[Twitter thread with @MicrosoftHelpt](https://twitter.com/adutchy76/status/1323711296085385216)

**Comment by Biswa96** [comment]
The `$(tty)` will results `/dev/ttyX` in WSL1 and `/dev/pts/X` in WSL2. It's by design. So, what does `GPG_TTY` expect?

**Comment by TobiasDeBruijn** [comment]
I think any TTY would work, regardless of where it is.

I'm getting /dev/ttyX on wsl.exe, or via the terminal app. So that seems to be all WSL1

**Comment by therealkenc** [comment]
It isn't related to WSL1 vs WSL2, but whatever is sending output to `/dev/tty1`, it is doing so by-design.

![image](https://user-images.githubusercontent.com/7049700/98759695-7de9db80-2386-11eb-91ed-442fbd3fdfd2.png)


The Linux penguin icon in the screencap looks somewhat suspect. Might or might not be related. To the extent you are launching Ubuntu from the Start menu, it would normally have an Ubuntu icon.

![image](https://user-images.githubusercontent.com/7049700/98759854-cb664880-2386-11eb-8ffd-badaed38099d.png)

Try the steps in the first screencap (with your WSL1 distro name of choice) and post the screencap. It is possible there was a behavioral change since 18363 but I don't have that build handy to check.

**Comment by TobiasDeBruijn** [comment]
That does seem to work. The screenshots:
![image](https://user-images.githubusercontent.com/28569170/98786555-8a117f80-23fe-11eb-988e-13c68d03bc94.png)

Whilst echoing the TTY's, I also echoed GPG_TTY, which did not match the TTY of the terminal.
![image](https://user-images.githubusercontent.com/28569170/98786643-ad3c2f00-23fe-11eb-9f34-e48b41fafc84.png)

Could this be because ~/.bashrc is only ran on the first time a terminal opens?

And the Pinguin icon is there when you open wsl.exe from the start menu, or shift-rightclick > Open Linux Terminal in Explorer

**Comment by therealkenc** [comment]
> shift-rightclick > Open Linux Terminal in Explorer

Okay.

> Could this be because ~/.bashrc is only ran on the first time a terminal opens?

Improbable. If `.bashrc` was not run then `GPG_TTY` would not be set at all. The spaces around "`=`" in the OP are incorrect though, if they are to be taken literally. With shift-rightclicky....

![image](https://user-images.githubusercontent.com/7049700/98790803-0e411400-23b9-11eb-8c11-10b03bee95f6.png)

Do the repro as suggested: Two powershell (or cmd) instances, `wsl.exe` from the prompt, like my previous post, with the two lines including the `echo` at the bottom of your `.bashrc`.

**Comment by TobiasDeBruijn** [comment]
I've added `echo $GPG_TTY` to ~/.bashrc. After doing this I closed the WSL window and launched it as you suggested via powershell. GPG_TTY was not echoed however. 
![image](https://user-images.githubusercontent.com/28569170/98967337-a7347400-250c-11eb-9e74-dbf8f29e1f0e.png)

After starting WSL a second time via PS, $(tty) and $GPG_TTY did not match: 
![image](https://user-images.githubusercontent.com/28569170/98967528-dba83000-250c-11eb-978b-bec8350827e7.png)

The spaces around the '=' you pointed out were fixed before this 'test'

**Comment by therealkenc** [comment]
You have a `tail -3` above (thanks!) but only two lines were returned. Which is at best unusual. There is likely something incorrect with your shell/login scripts, but I am unable to guess with the information provided. Your `GPG_TTY` is being set _somewhere_. 

One big-hammer approach is to install a clean distribution of another name (say Ubuntu-20.04 from the [store](https://www.microsoft.com/en-us/p/ubuntu-2004-lts/9n6svws3rx71#activetab=pivot:overviewtab)), put an `echo hello` at the top of `.bashrc`, and then move forward from there.

**Comment by TobiasDeBruijn** [comment]
Makes sense that it returned only 2 I think, since there were only 2 lines in the file :D
